### PR TITLE
Implemented NodeAffinity

### PR DESCRIPTION
**Pull request by ellistarn**
**1. Issue, if available:**
https://github.com/awslabs/karpenter/issues/482

**2. Description of changes:**
- Taint and Label generation will no longer replace existing taints/labels defined on the provisioner (layered constraints vs overrides)
- Preferences are currently respected as hard constraints (may be relaxed in the future)
- [aws] Capacity Type is now Capacity Types, prioritize spot if able. This will default to on-demand if unspecified.

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [X] Yes, issue opened: https://github.com/awslabs/karpenter/issues/704
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cloudprovider/aws/instance.go [hunk 1]
```diff
@@ -103,7 +104,14 @@ func (p *InstanceProvider) launchInstance(ctx context.Context,
 	launchTemplateName string,
 	instanceTypeOptions []cloudprovider.InstanceType,
 	subnets []*ec2.Subnet,
-	capacityType string) (*string, error) {
+	capacityTypes []string) (*string, error) {
+
+	// Select Spot if within constarints
```
Comment (JacobGabrielson) [inline pkg/cloudprovider/aws/instance.go:109]: Could you explain why we're doing this, in the comment? Right now the comment is just kinda saying the same thing.

**Review by JacobGabrielson — COMMENTED** [review]
Looking good so far... reviewed about 17/29 so far

File: pkg/utils/functional/suite_test.go [hunk 2]
```diff
@@ -82,4 +82,39 @@ var _ = Describe("Functional", func() {
...
+			Expect(IntersectStringSlice(nilset, universe)).To(Equal(universe))
+			Expect(IntersectStringSlice(universe, nilset)).To(Equal(universe))
+		})
+		Specify("empty set", func() {
+			Expect(IntersectStringSlice(empty, nilset)).To(BeEmpty())
+			Expect(IntersectStringSlice(nilset, empty)).To(BeEmpty())
+			Expect(IntersectStringSlice(universe, empty)).To(BeEmpty())
+		})
+		Specify("intersect", func() {
+			Expect(IntersectStringSlice(universe, subset)).To(ConsistOf("a", "b"))
```
Comment (JacobGabrielson) [inline pkg/utils/functional/suite_test.go:106]: why not `ConsistOf(subset)`?

File: pkg/utils/functional/suite_test.go [hunk 3]
```diff
@@ -82,4 +82,39 @@ var _ = Describe("Functional", func() {
...
+		disjoint := []string{"d", "e"}
+		duplicates := []string{"a", "a"}
+		Specify("nil set", func() {
+			Expect(IntersectStringSlice()).To(BeNil())
+			Expect(IntersectStringSlice(nilset)).To(BeNil())
+			Expect(IntersectStringSlice(nilset, nilset)).To(BeNil())
+			Expect(IntersectStringSlice(nilset, universe)).To(Equal(universe))
+			Expect(IntersectStringSlice(universe, nilset)).To(Equal(universe))
+		})
+		Specify("empty set", func() {
```
Comment (JacobGabrielson) [inline pkg/utils/functional/suite_test.go:100]: seems like should add a test with 3+ args, where the final arg is empty

File: pkg/utils/functional/suite_test.go [hunk 4]
```diff
@@ -82,4 +82,39 @@ var _ = Describe("Functional", func() {
...
 	})
+	Context("IntersectStringSlice", func() {
+		var nilset []string
+		empty := []string{}
+		universe := []string{"a", "b", "c"}
+		subset := []string{"a", "b"}
+		overlap := []string{"a", "b", "d"}
+		disjoint := []string{"d", "e"}
+		duplicates := []string{"a", "a"}
+		Specify("nil set", func() {
```
Comment (JacobGabrielson) [inline pkg/utils/functional/suite_test.go:93]: seems like should add at least one 3-arg test here

File: pkg/utils/functional/suite_test.go [hunk 5]
```diff
@@ -82,4 +82,39 @@ var _ = Describe("Functional", func() {
...
+		Specify("intersect", func() {
+			Expect(IntersectStringSlice(universe, subset)).To(ConsistOf("a", "b"))
+			Expect(IntersectStringSlice(subset, universe)).To(ConsistOf("a", "b"))
+			Expect(IntersectStringSlice(universe, overlap)).To(ConsistOf("a", "b"))
+			Expect(IntersectStringSlice(overlap, universe)).To(ConsistOf("a", "b"))
+			Expect(IntersectStringSlice(overlap, disjoint, universe)).To(BeEmpty())
+			Expect(IntersectStringSlice(universe, disjoint)).To(BeEmpty())
+			Expect(IntersectStringSlice(disjoint, universe)).To(BeEmpty())
+		})
+		Specify("duplicates", func() {
```
Comment (JacobGabrielson) [inline pkg/utils/functional/suite_test.go:114]: what about if the final arg is nil? (and there are at least 3 args?)

**Review by JacobGabrielson — COMMENTED** [review]
Looks good o'erall. A few minor comments.

File: pkg/cloudprovider/aws/cloudprovider.go [hunk 6]
```diff
@@ -196,3 +180,20 @@ func (c *CloudProvider) Default(ctx context.Context, v1alpha4constraints *v1alph
 		logging.FromContext(context.Background()).Errorf("failed to serialize provider, %s", err.Error())
 	}
 }
+
+// Constrain applies the pod's scheduling constraints to the constraints.
+// Returns an error if the constraints cannot be applied.
+func (c *CloudProvider) Constrain(ctx context.Context, v1alpha4constraints *v1alpha4.Constraints, pods ...*v1.Pod) error {
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/cloudprovider/aws/cloudprovider.go:186]: Thoughts on something better? I have a type collision because I'm aliasing the concept in this package.

File: pkg/cloudprovider/aws/instance.go (hunk 1 above)
Comment (ellistarn) [reply pkg/cloudprovider/aws/instance.go:109]: +1

File: pkg/scheduling/taints.go [hunk 7]
```diff
@@ -21,27 +21,28 @@ import (
...
 
-// Tolerates returns an error if the pod does not tolerate the taints
-// https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#concepts
-func Tolerates(pod *v1.Pod, taints ...v1.Taint) error {
-	var err error
-	for _, taint := range taints {
+type Taints []v1.Taint
+
+// Has returns true if taints has a taint for the given key
+func (ts Taints) Has(taint v1.Taint) bool {
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/scheduling/taints.go:27]: Yeah this is really eating away at me. I could start doing maps, but I think the conversion cost is just as expensive. Would love any advice on this. FWIW, I think a lot of this are 2-3 elements long.

File: pkg/controllers/allocation/scheduling/constraints.go [hunk 8]
```diff
@@ -15,56 +15,141 @@ limitations under the License.
...
+	return constraints, nil
 }
 
-func overrideTaints(taints []v1.Taint, pod *v1.Pod) []v1.Taint {
-	overrides := []v1.Taint{}
-	// Generate taints from pod tolerations
+func generateTaints(constraints *v1alpha4.Constraints, pod *v1.Pod) error {
+	taints := scheduling.Taints(constraints.Taints)
 	for _, toleration := range pod.Spec.Tolerations {
 		// Only OpEqual is supported
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/controllers/allocation/scheduling/constraints.go:56]: Updated

File: pkg/scheduling/taints.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (JacobGabrielson) [reply pkg/scheduling/taints.go:27]: Probably not worth it right now, mostly a note for the future. If it were an issue, potentially some kind of indexed list type data structure (a little difficult w/o generics I think).

**Review by JacobGabrielson — APPROVED** [review]

File: pkg/apis/provisioning/v1alpha4/register.go [hunk 9]
```diff
@@ -52,16 +52,19 @@ var (
...
-	ValidationHook            = func(ctx context.Context, constraints *Constraints) *apis.FieldError { return nil }
-	DefaultingHook            = func(ctx context.Context, constraints *Constraints) {}
+	// WellKnownLabels supported by karpenter and their allowable values
+	WellKnownLabels = map[string][]string{
+		v1.LabelArchStable:         {},
+		v1.LabelOSStable:           {},
+		v1.LabelTopologyZone:       {},
+		v1.LabelInstanceTypeStable: {},
+	}
+	DefaultHook = func(ctx context.Context, constraints *Constraints) {}
```
Comment (bwagner5) [inline pkg/apis/provisioning/v1alpha4/register.go:65]: could remove the parameter variables and just have types in the function signature.

File: pkg/cloudprovider/aws/apis/v1alpha1/constraints_validation.go [hunk 10]
```diff
@@ -19,55 +19,50 @@ import (
...
 	)
 }
 
-func (a *AWS) validateCapacityType(ctx context.Context) (errs *apis.FieldError) {
-	capacityTypes := []string{CapacityTypeSpot, CapacityTypeOnDemand}
-	if !functional.ContainsString(capacityTypes, aws.StringValue(a.CapacityType)) {
-		errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("%s not in %v", aws.StringValue(a.CapacityType), capacityTypes), "capacityType"))
-	}
-	return errs
+func (c *Constraints) validateCapacityType(ctx context.Context) (errs *apis.FieldError) {
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/apis/v1alpha1/constraints_validation.go:41]: validateCapacityTypes (plural) ?

File: pkg/cloudprovider/aws/instance.go [hunk 11]
```diff
@@ -103,30 +104,19 @@ func (p *InstanceProvider) launchInstance(ctx context.Context,
...
-			}
-		}
+	capacityTypes []string) (*string, error) {
+
+	// If unconstrained, default to spot to save on cost, otherwise use on
+	// demand. Constraint solving logic will guarantee that capacityTypes are
+	// either unconstrained (nil), or a valid WellKnownLabel. Provisioner
+	// defaulting logic will currently default to [on-demand] if unspecifed.
+	capacityType := v1alpha1.CapacityTypeSpot
+	if capacityTypes != nil && !functional.ContainsString(capacityTypes, v1alpha1.CapacityTypeSpot) {
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/instance.go:114]: interesting, I guess the ICE Cache could key on capacity-type too in order to support a native fallback to OD if we can't get spot capacity?

**Review by bwagner5 — COMMENTED** [review]
Looking good, I'm a little over half way through reviewing. But will have to look at the rest a bit later.

File: pkg/apis/provisioning/v1alpha4/register.go (hunk 9 above)
Comment (ellistarn) [reply pkg/apis/provisioning/v1alpha4/register.go:65]: I did this intentionally to cause the default behavior to no-op rather than be nil (and inevitably crash or require a nil check at the call site).

File: pkg/cloudprovider/aws/instance.go (hunk 11 above)
Comment (ellistarn) [reply pkg/cloudprovider/aws/instance.go:114]: Yeah totally.

**Review by bwagner5 — DISMISSED** [review]
A few small comments, but looks great!

File: pkg/controllers/allocation/scheduling/constraints.go [hunk 12]
```diff
@@ -15,56 +15,138 @@ limitations under the License.
...
+}
+
+func validateAffinity(pod *v1.Pod) (errs error) {
+	if pod.Spec.Affinity == nil {
+		return nil
+	}
+	if pod.Spec.Affinity.PodAffinity != nil {
+		errs = multierr.Append(errs, fmt.Errorf("pod affinity is not supported"))
+	}
+	if pod.Spec.Affinity.PodAntiAffinity != nil {
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/controllers/allocation/scheduling/constraints.go:124]: Until we respect the constraints, we basically *have* to not schedule the pods. I can imagine eventually trying to implement pod affinity, but minimally it will be incremental effort over this. FWIW, the current behavior is to not provision if constraints are not supported.

**Review by bwagner5 — APPROVED** [review]
lgtm

**Review by JacobGabrielson — APPROVED** [review]

**Review by bwagner5 — APPROVED** [review]
lgtm
